### PR TITLE
HIR-20. Add promotional offers management endpoints

### PR DESCRIPTION
**Pull request by oleg1std**

### CODE REVIEW

File: HappyTravel.Hiroshima.DirectManager/Infrastructure/Extensions/DbContextExtensions.cs [hunk 1]
```diff
@@ -0,0 +1,60 @@
+﻿using System.Collections.Generic;
+using System.Linq;
+using System.Threading.Tasks;
+using CSharpFunctionalExtensions;
+using HappyTravel.Hiroshima.Data;
+using Microsoft.EntityFrameworkCore;
+
+namespace HappyTravel.Hiroshima.DirectManager.Infrastructure.Extensions
+{
+    public static class DbContextExtensions
```
Comment (kirillta) [inline HappyTravel.Hiroshima.DirectManager/Infrastructure/Extensions/DbContextExtensions.cs:10]: Those methods evaluate queries for sure. Don't know if we may call that _context_ extensions
@mingazhev

File: HappyTravel.Hiroshima.DirectManager/Services/PromotionalOfferManagementService.cs [hunk 2]
```diff
@@ -0,0 +1,134 @@
...
+                return;
+
+            _dbContext.RoomPromotionalOffers.RemoveRange(promotionalOffers);
+            await _dbContext.SaveChangesAsync();
+
+            _dbContext.DetachEntries(promotionalOffers);
+        }
+        
+        
+        private async Task<Result<List<RoomPromotionalOffer>>> CheckAndGetPromotionalOffersToRemove(int contractId, int contractManagerId,
```
Comment (kirillta) [inline HappyTravel.Hiroshima.DirectManager/Services/PromotionalOfferManagementService.cs:67]: No need to enforce the _"check"_ part. One expects significant conditions from a method with a name like that. I suggest simply `GetPromotionalOffersToRemove` here

File: HappyTravel.Hiroshima.DirectManager/Models/Responses/PromotionalOffer.cs [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+        }
+
+
+        public int Id { get; }
+        public int ContractId { get; }
+        public int RoomId { get; }
+        public DateTime BookBy { get; }
+        public DateTime ValidFrom { get; }
+        public DateTime ValidTo { get; }
+        public double DiscountPercent { get; }
```
[the remark this answers is not in the document]
Comment (oleg1std) [reply HappyTravel.Hiroshima.DirectManager/Models/Responses/PromotionalOffer.cs:28]: I haven't seen any other types of discount in available contracts

[the remark this answers is not in the document]
Comment (oleg1std) [reply HappyTravel.Hiroshima.DirectManager/Models/Responses/PromotionalOffer.cs:28]: Yes, I've seen this. Setting percent, as I see, is a common hotel's way to set a discount, and it is used in the available direct contracts. Our new format is about interservice interaction

File: HappyTravel.Hiroshima.DirectManager/Infrastructure/Extensions/DbContextExtensions.cs (hunk 1 above)
Comment (oleg1std) [reply HappyTravel.Hiroshima.DirectManager/Infrastructure/Extensions/DbContextExtensions.cs:10]: Ok. I'll rename it to ContractManagementQueryExtensions

File: HappyTravel.Hiroshima.DirectManager/Models/Responses/PromotionalOffer.cs (hunk 3 above)
[the remark this answers is not in the document]
Comment (kirillta) [reply HappyTravel.Hiroshima.DirectManager/Models/Responses/PromotionalOffer.cs:28]: Yes, sure. Just keep in mind other discount types and formats may appear in the future

**Pull request merged**